core/src/main/scala/kafka/server/BrokerRaftEventManager.scala

```diff
@@ -0,0 +1,165 @@
+}
+
+case object ShutdownEventThread extends RaftMessage {
+  override def correlationId(): Int = -1
+  override def data(): ApiMessage = null
+}
+
+class BrokerRaftEventManager(replicaManager: ReplicaManager,
+                             groupCoordinator: GroupCoordinator,
+                             metadataCache: MetadataCache,
+                             quotas: QuotaManagers,
+                             clusterId: String,
+                             time: Time,
+                             optionalProcessor: Option[RaftEventProcessor] = None,
+                             eventQueueTimeTimeoutMs: Long = 300000) extends KafkaMetricsGroup with RaftEventProcessor {
+
+  import BrokerRaftEventManager._
+
+  val queue = new LinkedBlockingQueue[QueuedEvent]
+
+  val thread = new BrokerRaftEventThread(BrokerRaftEventThreadName)
+
+  val processor: RaftEventProcessor = optionalProcessor.getOrElse(this)
+
+  // metrics
+  private val eventQueueTimeHist = newHistogram(EventQueueTimeMetricName)
+  newGauge(EventQueueSizeMetricName, () => queue.size)
+
+  def start(): Unit = thread.start()
+
+  def close(): Unit = {
+    try {
+      thread.initiateShutdown()
+      thread.awaitShutdown()
+    } finally {
+      removeMetric(EventQueueTimeMetricName)
+      removeMetric(EventQueueSizeMetricName)
+    }
+  }
+
+  def put(event: RaftMessage): QueuedEvent = {
+    val queuedEvent = new QueuedEvent(event, time.milliseconds())
+    queue.put(queuedEvent)
+    queuedEvent
+  }
+
+  class BrokerRaftEventThread(name: String) extends ShutdownableThread(name = name, isInterruptible = false) {
+    logIdent = s"[BrokerRaftEventThread] "
+
+    override def doWork(): Unit = {
+      val dequeued = pollFromEventQueue()
+      dequeued.event match {
+        case ShutdownEventThread => // The shutting down of the thread has been initiated at this point. Ignore this event.
+        case StartupBeforeAnyProcessing => processor.startupBeforeAnyProcessing()
+        case raftMessage =>
+          eventQueueTimeHist.update(time.milliseconds() - dequeued.enqueueTimeMs)
+          try {
+            processor.process(raftMessage)
+          } catch {
+            case e: Throwable => error(s"Uncaught error processing event $raftMessage", e)
+          }
+      }
+    }
+  }
+
+  private def pollFromEventQueue(): QueuedEvent = {
+    val hasRecordedValue = eventQueueTimeHist.count() > 0
+    if (hasRecordedValue) {
+      val event = queue.poll(eventQueueTimeTimeoutMs, TimeUnit.MILLISECONDS)
+      if (event == null) {
+        eventQueueTimeHist.clear()
+        queue.take()
+      } else {
+        event
+      }
+    } else {
+      queue.take()
+    }
+  }
+
+  override def startupBeforeAnyProcessing(): Unit = {
+  }
+
+  override def process(event: RaftMessage): Unit = {
+    try {
+      trace(s"Handling event:$event")
+      val data = event.data
+      data.apiKey match {
+        case ApiKeys.UPDATE_METADATA.id => handleUpdateMetadataRequest(event)
```

Comment:
`ApiKeys.UPDATE_METADATA` identifies an RPC message, but this is not an RPC, of course.

Check out the classes in `metadata/src/main/resources/common/metadata/`.  Those are the ones that you'll see here.